blockchain/blockchain.go

```diff
@@ -1824,6 +1824,11 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
 
 		cache, _ := bc.stateCache.TrieDB().Size()
 		stats.report(chain, i, cache)
+
+		// update governance CurrentSet if it is at an epoch block
+		if bc.engine.CreateSnapshot(bc, block.NumberU64(), block.Hash(), nil) != nil {
+			return i, events, coalescedLogs, err
+		}
```

Comment:
How can I know that this lines of code only makes a new snapshot at an `epoch block`?

Reply:
Umm.. `snapshot` function queries a snapshot for a block, and if needed, it creates a snapshot for a block.